### PR TITLE
Fixed #24742 -- Made runserver.check_migrations ignore read-only databases

### PR DESCRIPTION
**Pull request by claudep**
Thanks Luis Del Giudice for the report.

### CODE REVIEW

**Comment by delgiudices** [comment]
Thanks for helping with it @claudep

**Comment by aaugustin** [comment]
Looks good to me.

**Comment by MarkusH** [comment]
I'm not sure this is the approach I'd use to fix the issue. As already pointed out by @aaugustin on trac #24669 is somewhat related. I think we should fix that issue and instead of making `runserver` fail with readonly databases, issue a message that the user is working on a readonly database or something.

**Comment by claudep** [comment]
Markus, this patch doesn't make `runserver` fail with readonly database. It may not be clear while reading the patch, but the `return` statement returns from the internal `check_migrations` method, that is the `runserver` command will simply continue. We could still add an extra message if needed.

**Comment by MarkusH** [comment]
Oh, I should rephrase my sentence I guess.

Right now `runserver` fails when accessing a readonly (default) database. This is due to a call to `django.db.migrations.recorder.MigrationRecorder.ensure_schema()` inside `django.db.migrations.recorder.MigrationRecorder.applied_migrations()` inside `django.db.migrations.loader.MigrationLoader.build_graph()`.

Instead of catching an arbitrary `DatabaseError` somewhere in `runserver` (what about missing read permissions on the migrations table when the schema already exists?) I much rather raise an explicit `MigrationSchemaMissing` exception in the `MigrationRecorder` and catch that wherever needed.

Having the development server silently continuing to work may not be the best idea. This could lead to interesting errors. We should at least issue a warning to the user.

**Comment by claudep** [comment]
I tried to improve the patch (purposefully in a separate commit) a little. Markus, is that the direction you pointed?

**Comment by MarkusH** [comment]
Thanks @claudep, that's what I had in mind.

Can you reorder the commits and first raise the exception and issue a warning in the second commit, if that makes sense?

File: django/db/migrations/recorder.py [hunk 1]
```diff
@@ -2,10 +2,15 @@
 
 from django.apps.registry import Apps
 from django.db import models
+from django.db.utils import DatabaseError
 from django.utils.encoding import python_2_unicode_compatible
 from django.utils.timezone import now
 
 
+class MigrationSchemaMissing(DatabaseError):
```
Comment (MarkusH) [inline django/db/migrations/recorder.py]: Can you move the exception to `django.db.migrations.exceptions`, please.

Comment (claudep) [inline django/db/migrations/recorder.py]: Oh, of course!

**Comment by claudep** [comment]
I intend to merge both commits in the final patch.

**Comment by MarkusH** [comment]
Not sure if you want to mention it in the 1.9 release notes for the changes to management commands. Otherwise, LGTM

**Comment by timgraham** [comment]
LGTM, I don't think release notes are needed.

**Pull request merged**